File: Search/binary_search_iterative.py
#O(logN),N is the length of the array
def binary_search_iterative(array,target,start,end):
    while start <= end:
        mid = (start + end) // 2
        #if target, return mid
        if array[mid] == target:
            return mid
        #if target is smaller than mid
        elif array[mid] > target:
            end = mid - 1
        #if target is larger than mid
        else:
            start = mid + 1
    return None

array = [1,2,3,4,5,6,7,8,9,10]
print(binary_search_iterative(array,8,0,len(array)-1))

File: Search/binary_search_iterative.java
import java.util.*;

//In order to use binary search, we should assume the array is sorted.
// binary search (recursive)
// runtime: O(log n)
public class binary_search_iterative {
  public static int binarySearch(int[] arr, int target, int start, int end) {
    while (start <= end) {
      int mid = (start + end) / 2;
      if (arr[mid] == target) {
        return mid;
      }
      else if (arr[mid] > target) {
        end = mid - 1;
      }
      else {
        start = mid + 1;
      }
    }
    return -1;
  }
}
